internal/datastore/postgres/postgres.go

```diff
@@ -123,22 +124,45 @@ func NewPostgresDatastore(
 }
 
 func newPostgresDatastore(
-	url string,
+	pgURL string,
 	options ...Option,
 ) (datastore.Datastore, error) {
 	config, err := generateConfig(options)
 	if err != nil {
 		return nil, fmt.Errorf(errUnableToInstantiate, err)
 	}
 
+	// check for plan_cache_mode and, if not specified, set to `force_custom_plan`.
+	// See: https://spicedb.dev/d/force-custom-plan for more information.
+	parsed, err := pgxpool.ParseConfig(pgURL)
+	if err != nil {
+		return nil, fmt.Errorf(errUnableToInstantiate, err)
+	}
+
+	if existing, ok := parsed.ConnConfig.Config.RuntimeParams["plan_cache_mode"]; ok {
+		log.Info().Str("plan_cache_mode", existing).Msg("found plan_cache_mode in DB URL; leaving as-is")
+	} else {
+		parsedURL, err := url.Parse(pgURL)
```

Comment:
Why do we have to handle this case? Does pgx not handle both URI types?

Reply:
PGX does, but we don't and PGX doesn't provide a way that i could find to modify the URL

Reply:
why are we even modifying the URL when we could just wrap this logic up into the config parser:

```go
func ParseConfig(connString string) (*ConnConfig, error) {
	config, err := pgxpool.ParseConfig(connString)
	if err != nil {
		return config, err
	}
	
	if existing, ok := config.ConnConfig.Config.RuntimeParams["plan_cache_mode"]; ok {
		log.Info().Debug("plan_cache_mode", existing).Msg("found plan_cache_mode in DB URI; leaving as-is")
	} else {
		config.ConnConfig.Config.RuntimeParams["plan_cache_mode"] = "force_custom_plan"
		log.Warn().
			Str("details-url", "https://spicedb.dev/d/force-custom-plan").
			Str("plan_cache_mode", "force_custom_plan").
			Msg("defaulting value in Postgres DB URI") // this wording in case we add more defaults in the future
	}

	return config, nil
}

readPoolConfig, err := ParseConfig(pgURL)
...
writePoolConfig, err := ParseConfig(pgURL)


```

Reply:
I'm not certain that PGX passes `RuntimeParams` modifications to Postgres; I read through the code and didn't see it do so (unless I missed it). That's why I changed it to modify the URL

Reply:
it does, here: https://github.com/jackc/pgx/blob/70a200cff4d42098786c3649b2f3cd30b7a66f1b/pgconn/pgconn.go#L299-L302

so let's go with something along these lines:

```go
// defaultCustomPlan parses a Postgres URI and determines if a plan_cache_mode
// has been specified. If not, it defaults to "force_custom_plan".
// This worksaround a bug impacting performance documented here:
// https://spicedb.dev/d/force-custom-plan for more information.
func defaultCustomPlan(config *pgx.ConnConfig, err error) (*pgx.ConnConfig, error) {
	if err != nil {
		return config, err
	}

	if existing, ok := config.ConnConfig.Config.RuntimeParams["plan_cache_mode"]; ok {
		log.Info().
			Debug("plan_cache_mode", existing).
			Msg("found plan_cache_mode in DB URI; leaving as-is")
		return
	}

	config.ConnConfig.Config.RuntimeParams["plan_cache_mode"] = "force_custom_plan"
	log.Warn().
		Str("details-url", "https://spicedb.dev/d/force-custom-plan").
		Str("plan_cache_mode", "force_custom_plan").
		Msg("defaulting value in Postgres DB URI")

	return config, nil
}

...

pgConfig, err := defaultCustomPlan(pgxpool.ParseConfig(pgURL))
if err != nil {
	return nil, fmt.Errorf(errUnableToInstantiate, err)
}


readPoolConfig, err := pgConfig.Copy()
...
writePoolConfig, err := pgConfig.Copy()
```